jv-online-book-store

ipantazi opened the pull request:
Removed CommandLineRunner bean from `Application` class.
Created classes: `BookController`, `BookDto`, `CreateBookRequestDto`, `EntityNotFoundException`, `MapperConfig` and interface `BookMapper`.
Added method `getBookById(Long id)` to classes: `BookRepositoryImpl`, `BookServiceImpl` and interface `BookRepository`.
Added dependencies to `pom.xml`.

Added to `application.properties` class - `spring.jpa.open-in-view=false` because was `WARN 7132 spring.jpa.open-in-view is enabled by default`.
Added to `BookMapper` interface - `@Mapping(target = "id", ignore = true)` because was `[WARNING] Unmapped target property: "id"`.

ipantazi committed:
Added `Book` entity, class `DataProcessingException`, `BookRepository` interface with methods and `BookRepositoryImpl` class.
Added `BookService` interface with methods and `BookServiceImpl` class.
Modified class `Application` and file `application.properties`.
Created a new folder `resources` in the `src/test` directory
Create a new file `application.properties` in the `src/test/resources` folder.
Added dependencies to the `pom.xml` file.

ipantazi committed:
Using EntityManagerFactory instead of SessionFactory in `BookRepositoryImpl` class

ipantazi committed:
Deleted dependency `commons-dbcp` in `pom.xml`.
Set upped all fields and renamed object in class `Application`.
 Renamed schema in `application.properties`.
 Added annotation `+@RequiredArgsConstructor` and deleted constructor in classes `BookServiceImpl` and `BookRepositoryImpl`.
 Added required constraints for fields, changed annotation `@Data` to `@Getter`, `@Setter`, `@Builder`, `@ToString`, `@NoArgsConstructor`, `@AllArgsConstructor`, added annotation `@Table` and methods `equals` and `hashCode` in class `Book`.

ipantazi committed:
Removed methods: `equals()`, hashCode();
annotations: `@ToString`, `@NoArgsConstructor`, `@AllArgsConstructor`, `@Builder`;
 modified annotations `@Column` for all fields in `Book` class.

ipantazi committed:
Removed CommandLineRunner bean from `Application` class.
Created classes: `BookController`, `BookDto`, `CreateBookRequestDto`, `EntityNotFoundException`, `MapperConfig` and interface `BookMapper`.
Added method `getBookById(Long id)` to classes: `BookRepositoryImpl`, `BookServiceImpl` and interface `BookRepository`.
Added dependencies to `pom.xml`.

ipantazi committed:
Added to `application.properties` class - `spring.jpa.open-in-view=false` because was `WARN 7132 spring.jpa.open-in-view is enabled by default`.
Added to `BookMapper` interface - `@Mapping(target = "id", ignore = true)` because was `[WARNING] Unmapped target property: "id"`.

ipantazi committed:
Added version lombok to the dependency in the `pom.xml`.

ipantazi committed:
Modified dependency lombok in the `pom.xml`.

JJJazl commented on src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.Optional;
+import mate.academy.onlinebookstore.model.Book;
+
+public interface BookRepository {
```
```suggestion
public interface BookRepository extends JpaRepository<Book, Long> {
```

JJJazl commented on src/main/java/mate/academy/onlinebookstore/repository/BookRepositoryImpl.java:
```diff
@@ -0,0 +1,62 @@
+        try {
+            entityManager = entityManagerFactory.createEntityManager();
+            transaction = entityManager.getTransaction();
+            transaction.begin();
+            entityManager.persist(book);
+            transaction.commit();
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+                throw new DataProcessingException("Can't insert book to DB: " + book, e);
+            }
+        } finally {
+            if (entityManager != null) {
+                entityManager.close();
+            }
+        }
+        return book;
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("from Book", Book.class).getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't find all books", e);
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return Optional.ofNullable(entityManager
+                    .createQuery("from Book where id = :id", Book.class)
+                    .setParameter("id", id)
+                    .getSingleResult());
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't get book by id: " + id, e);
+        }
+    }
+}
```
I think you need to use a functionality of `spring-data-jpa` dependency, not a raw JPA/Hibernate API
```suggestion
```

JJJazl commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,40 @@
+@Getter
+@Setter
+public class Book {
+    @Column(nullable = false)
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
You don't need this annotation for `id` field, because value for this field is generated by the database when Hibernate stores your entity. I suggest to read about `GenerationType` strategies in Hibernate
```suggestion
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
```

ipantazi committed:
Removed the annotation `@Column(nullable = false)` from the class `Book`.
The interface `BookRepository` is extended by the interface `JpaRepository`, all methods from this interface are removed and the class `BookRepositoryImpl` is removed.

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toBookDto)
+                .collect(Collectors.toList());
```
```suggestion
                .toList();
```

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/controller/BookController.java:
```diff
@@ -0,0 +1,35 @@
+        return bookService.findById(id);
+    }
+
+    @PostMapping
```
ResponseStatus.CREATED

liliia-ponomarenko reviewed (approved):
Great! A few minor comments ;)
